### ISSUE TITLE
False positive mixed - failure to properly infer type

### ISSUE THREAD

**Issue by sserbin** [vimeo/psalm#1122]
```php
<?php
function foo() : DateTime {
    if (rand(0, 1)) {
    	$dt = new DateTime;
    }
    if (!isset($dt)) {
    	$dt = new DateTime;
    }
  
    strlen($dt);
    return $dt;
}
```
https://getpsalm.org/r/23088a5b14

Expected: `$dt` is `DateTime`
Actual: `$dt` is mixed

**Comment by muglug** [comment]
This was broken in 3.0.3. For the moment, you can fix by initialising `$dt` like so: https://getpsalm.org/r/f00f1bb67b

**Comment by muglug** [comment]
The long term solution is for Psalm to track types of possibly undefined variables.
